crmsh/bootstrap.py

```diff
@@ -828,24 +832,50 @@ def _parse_user_at_host(s: str, default_user: str) -> typing.Tuple[str, str]:
 
 def init_ssh():
     user_host_list = [_parse_user_at_host(x, _context.current_user) for x in _context.user_at_node_list]
-    init_ssh_impl(_context.current_user, user_host_list)
+    init_ssh_impl(
+        _context.current_user,
+        ssh_key.KeyFile(_context.ssh_key_file) if _context.ssh_key_file is not None else None,
+        user_host_list,
+    )
     if user_host_list:
+        service_manager = ServiceManager()
         for user, node in user_host_list:
-            if utils.service_is_active("pacemaker.service", remote_addr=node):
+            if service_manager.service_is_active("pacemaker.service", remote_addr=node):
                 utils.fatal("Cluster is currently active on {} - can't run".format(node))
 
 
-def init_ssh_impl(local_user: str, user_node_list: typing.List[typing.Tuple[str, str]]):
-    """ Configure passwordless SSH."""
-    utils.start_service("sshd.service", enable=True)
-    configure_ssh_key(local_user)
+def init_ssh_impl(local_user: str, ssh_public_key: typing.Optional[ssh_key.Key], user_node_list: typing.List[typing.Tuple[str, str]]):
+    """ Configure passwordless SSH.
+
+    The local_user on local host will be configured.
+    If user_node_list is not empty, those user and host will also be configured.
+    If ssh_public_key is specified, it will be added to authorized_keys; if not, a new key pair will be generated for each node.
+    """
+    ServiceManager(sh.LocalOnlyAutoShell(sh.LocalShell())).start_service("sshd.service", enable=True)
```

Comment:
I noticed that `ServiceManager(sh.LocalOnlyAutoShell(sh.LocalShell()))` is used in several places. Would it be worthwhile to simplify/wrapper it?

Reply:
For now, the usage of `LocalOnlyAutoShell` is quite limited. It is only used by `ServiceManager` during the early stage of bootstrap.

I prefer not to create a wrapper for it, or the wrapper function may look more confusing.